### ISSUE TITLE
Kiali graph showing istio-system resources in unrelated and selected namespace

### ISSUE THREAD

**Issue by jbilliau5668** [kiali/kiali#2967]
**Describe the bug**
I install a clean, brand new install of Kiali v1.20 via the operator. I install a test application that's exposed externally via the istio ingressgateway. I sent traffic to it, then go into Kiali to view the data. Kiali is showing me grafana.istio-system.svc.cluster.local and kiali.istio-system.svc.cluster.local even though I have selected "whale" as my namespace. If I select "istio-system" as well, it shows what is likely the real issue: Kiali thinks those services are in the whale namespace, which they definitely aren't.

**Versions used**
Kiali: 1.20
Istio: 1.5.4 - 1.6.4
Kubernetes flavour and version: AWS EKS 1.16.8

**To Reproduce**
1.) Install Istio 1.5.4 through 1.6.4.
2.) Install Kiali 1.20 via the operator.
3.) Send traffic to service in service mesh
4.) Go into Kiali and select namespace for service you just sent traffic to.

**Expected behavior**
I shouldn't see istio-system resources as being in the namespace I selected; I should see just the istio-ingressgateway going to my test-whale service. If I install Kiali via the built-in Istio method with 1.5.4, it works fine and as expected. ONLY when going to Kiali 1.20 as a standalone, operator-initiated install does this behavior happen.

![image](https://user-images.githubusercontent.com/59461284/86620341-28419480-bf8a-11ea-82ad-024f906332f2.png)

**Comment by jshaughn** [comment]
@jbilliau5668 I think for us to better understand this issue we're going to need to look at the prometheus telemetry that generates the unwanted nodes.  You can launch prometheus via 

`> istioctl dashboard prometheus`

From there enter the following query:

`istio_requests_total{reporter="source",destination_service="kiali.istio-system.svc.cluster.local"}`

`istio_requests_total{reporter="source",destination_service="grafana.istio-system.svc.cluster.local"}`

This should help us understand what Istio is feeding into Kiali, and whether the issue is in the Istio telemetry or the Kiali graph.

**Comment by jbilliau5668** [comment]
hmmm good call, I assume this is the problem?


istio_requests_total{app="istio-ingressgateway",chart="gateways",connection_security_policy="unknown",destination_app=**"test-whale"**,destination_canonical_revision="1",destination_canonical_service="test-whale",destination_principal="unknown",destination_service="kiali.istio-system.svc.cluster.local",destination_service_name="**kiali.istio-system.svc.cluster.local"**,destination_service_namespace="whale",destination_version="1",destination_workload="test-whale",destination_workload_namespace="whale",heritage="Tiller",instance="172.18.42.177:15090",istio="ingressgateway",job="envoy-stats",namespace="istio-system",pod_name="istio-ingressgateway-bd646c88b-2dklw",pod_template_hash="bd646c88b",release="istio",reporter="source",request_protocol="http",response_code="500",response_flags="-",service_istio_io_canonical_name="istio-ingressgateway",service_istio_io_canonical_revision="latest",source_app="istio-ingressgateway",source_canonical_revision="latest",source_canonical_service="istio-ingressgateway",source_principal="unknown",source_version="unknown",source_workload="istio-ingressgateway",source_workload_namespace="istio-system"}

no idea how the hell that even happens....we did a fresh, brand new install of the Prometheus chart as well; looks like I need to start there?

**Comment by jmazzitelli** [comment]
That telemetry is the problem. Somehow Istio thinks your apps are behind the kiali service (???). No idea how that happened, but  Kiali is graphing the "correct thing" - the telemetry is just incorrect (i.e. garbage in - garbage out :)

**Comment by jbilliau5668** [comment]
hmm ok. thanks for the help, closing.

**Comment by jbilliau5668** [comment]
actually, I think this might be Kiali related. I installed Istio 1.6.3 with built-in prometheus, default profile. Deploy sample app, all metrics look fine; if I look at prometheus dashboard at `istio_requests_total`, they all look correct.

So then I turn Kiali on in my default.yaml manifest, and re-run "istioctl manifest apply -f default.yaml". Go to my exposed URL a few times, then go to prometheus....still looks good.

Now here's the kicker.....I then go to my dashboard URL (eksdashboard.xxxx.xxxx.xxxxx/kiali) and **initially**, the graph looks good. Within 10 seconds, kiali.istio-system.svc.cluster.local shows up. i go back to Prometheus dashboard and the metrics are infected! JUST by going to the Kiali URL this happens.....I have done nothing else except that between when the metrics were correct and when they shows this `destination_service` of kiali.istio-system.svc.cluster.local.

I know this sounds nuts, but I can't think of anything else it might be; i've uninstalled and reinstalled and retested at least a dozen times; going to the Kiali dashboard is what does it.

**Comment by jbilliau5668** [comment]
Also, all of this works fine in Istio 1.5.4 with bundles Kiali 1.15. Only in 1.6.X with 1.18 does this behavior happen, as well as the standalone chart of 1.20

**Comment by jmazzitelli** [comment]
@jbilliau5668 Check your Kiali pod and see if you have an Istio sidecar container injected into it. I suspect you have something like auto-injection of sidecars in the istio-system namespace (or wherever you put Kiali) and the Kiali pod is getting injected with an istio sidecar which then starts populating telemetry. I don't ever recall Kiali pod getting injected with an istio sidecar before, and AFAIK it should not be getting one.

**Comment by jbilliau5668** [comment]
nope, no sidecars on anything in my istio-system namespace. Upon further testing, this looks isolated to Istio 1.6.X. I just freshly installed 1.5.4 with Kiali set to `false`, and then installed the newest Kiali release via the operator=v.1.20.0. Everything work fine.

I'm not sure whose at fault here, Istio or Kiali, but no version of Kiali properly works with any version of Istio 1.6.X that I can see.

**Comment by jmazzitelli** [comment]
Telemetry did change in 1.6 - there was a bug that was recently fixed that was causing bad Kiali graphs but this is probably unrelated (it should have been fixed in the l latest 1.6.x release, too): https://github.com/istio/istio/issues/24379

Also, AFAIK, no one else has seen this behavior on Istio 1.6. So we probably need to figure out if you install is somehow different.

**Comment by jbilliau5668** [comment]
That's what puzzling; at first I thought it had to be me. There's no way this evaded everyone else using/testing Istio and Kiali right? That's why I ditched my custom manifest and just installed Istio with the default profile, Kiali enabled. It's as boilerplate and cookie-cutter as I can get it, and yet the moment I go into the Kiali console all my metrics get fubared. 

Not sure where to go from here. \_(ツ)_/¯

**Comment by jmazzitelli** [comment]
@jbilliau5668 can you provide the "exact" steps you do to install Istio and Kiali when you see the problem? Show the actual commands you run and link to the exact Istio distro you install - that would help someone attempt to recreate the problem.

**Comment by jmazzitelli** [comment]
@jbilliau5668 one other thing you can try - install the bookinfo demo (rather than your "whale" app) and see if you get the same behavior.

**Comment by jbilliau5668** [comment]
Sure.

1.) Downloaded the Istio 1.6.4 release. Navigate to bin folder and do a "_istioctl profile dump default > default.yaml_"

2.) I then edit this file to turn kiali and tracing on, as well as change ingress-gateway from `LoadBalancer` to `NodePort`, 
since we throw AWS ALB's in front of our ingress-gateways (we on AWS EKS). Then I run "_istioctl manifest apply -f default.yaml_" and off it goes, installing fine.

3.) I deploy a virtual service that routes eksdashboard.jasoneks.foc.zone to grafana, and eksdashboard.jasoneks.foc.zone/kiali to kiali.

4.) Ok, Istio installed. I then install sentiment analysis from here: [](https://github.com/rinormaloku/istio-mastery), exposing it externally as sa.jasoneks.foc.zone. So at this point I have a fresh cluster, Istio 1.6.4 installed in istio-system using slightly modified default profile, and sentiment-analysis in the sentiment-analysis namespace, exposed externally using istio-ingressgateway as an ingress.

5.) I navigate to the URL, and hit refresh a couple times. Then I do a "istioctl dashboard prometheus" and look at `istio_requests_total`. I see this:

![image](https://user-images.githubusercontent.com/59461284/86787736-6941b380-c033-11ea-99bd-79876e6fdba1.png)

No Kiali references, all looks good. Now, let me navigate to the Kiali dashboard.

6.) Initial graph looks good!:

![image](https://user-images.githubusercontent.com/59461284/86788056-bde52e80-c033-11ea-8bf9-7f7bce775be1.png)

7.) 10 seconds later:

![image](https://user-images.githubusercontent.com/59461284/86788089-c76e9680-c033-11ea-897e-4066249891bb.png)

![image](https://user-images.githubusercontent.com/59461284/86788154-d81f0c80-c033-11ea-9b7c-4f916103ca2e.png)

8.) That's about it.

**Comment by jmazzitelli** [comment]
> 3.) I deploy a virtual service that routes eksdashboard.jasoneks.foc.zone to grafana, and eksdashboard.jasoneks.foc.zone/kiali to kiali.

OK, what do those Virtual Services look like (the YAML).

**Comment by jbilliau5668** [comment]
This gets deployed after the istio install to expose Grafana, Kiali, and Jaeger:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: istio-dashboard
  namespace: istio-system
  annotations:
    zalando.org/aws-load-balancer-scheme: internet-facing
    kubernetes.io/ingress.class: alb-istio
  labels:
    app: istio-dashboard
spec:
  rules:
  - host: ${hostname}
    http:
      paths:
      - backend:
          serviceName: istio-ingressgateway
          servicePort: 80

---

apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: dashboard-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "${hostname}"

---

apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: dashboard-router
  namespace: istio-system
spec:
  gateways:
    - dashboard-gateway
  hosts:
    - "${hostname}"
  http:
  - match:
    - uri:
        prefix: /jaeger
    route:
    - destination:
        host: tracing.istio-system.svc.cluster.local
  - match:
    - uri:
        prefix: /kiali
    route:
    - destination:
        host: kiali.istio-system.svc.cluster.local
  - match:
    - uri:
        prefix: /
    route:
    - destination:
        host: grafana.istio-system.svc.cluster.local
```

**Comment by jmazzitelli** [comment]
My suspicion is either that yaml is wrong in some way OR there is an istio bug here. I'm almost sure this isn't a Kiali issue - Kiali is graphing the telemetry as found in Prometheus and Kiali isn't routing requests to itself. So something is wrong in the Istio routing. I posted something to Doug Reid in Istio slack's [`#extensions-telemetry`](https://istio.slack.com/archives/C382V8Q92) room to see if he has any thoughts.

**Comment by jbilliau5668** [comment]
Yeah I hear you; it's odd that ONLY upon loading the Kiali graph does this behavior happen. As for the YAML, it's the same YAML we've used since Istio 1.3; just routes to different paths based off uri, and works fine in Istio 1.5.4. It looks like the ingress-gateway itself is generating the Prometheus metrics that shows it going to kiali.istio-system.svc.cluster.local, but in the sentiment-analysis namespace, which makes no sense.


`istio_requests_total{connection_security_policy="unknown",destination_app="unknown",destination_canonical_revision="latest",destination_canonical_service="unknown",destination_principal="unknown",destination_service="kiali.istio-system.svc.cluster.local",destination_service_name="kiali.istio-system.svc.cluster.local",destination_service_namespace="unknown",destination_version="unknown",destination_workload="unknown",destination_workload_namespace="unknown",instance="172.18.67.148:15090",job="envoy-stats",namespace="istio-system",pod_name="istio-ingressgateway-74c68446bf-kpnqh",reporter="source",request_protocol="http",response_code="200",response_flags="-",source_app="istio-ingressgateway",source_canonical_revision="latest",source_canonical_service="istio-ingressgateway",source_principal="unknown",source_version="unknown",source_workload="istio-ingressgateway",source_workload_namespace="istio-system"}`
--

**Comment by jmazzitelli** [comment]
> it's odd that ONLY upon loading the Kiali graph

Remember that when you "load the Kiali graph" that's when you first start going into the Istio Gateway through the Virtual Service into Kiali - so it makes sense that that is when you start seeing that kiali telemetry show up in Prometheus - because it is only at that time that you started hitting the Kiali endpoint through the Gateway and the proxy is capturing that telemetry at that point.

I just don't understand why the destination_workload is "sa-frontend" and its namespace is "sentiment-analysis" - that's completely unrelated to Kiali :) It smells like some kind of Istio bug here that is capturing the wrong telemetry when you hit the Istio Gateway and go to Kiali through your Virtual Service.

**Comment by jbilliau5668** [comment]
Agreed. Would you like me to close this issue (again :) )? At this point, it looks like this should be an issue on the istio github if anything.

**Comment by jmazzitelli** [comment]
You might want to post the yaml for all the gateways/virtual services for the sentiment-analysis stuff too. I wonder if some "wires are getting crossed" - I haven't looked at what Istio changed between 1.5 and 1.6 wrt Gateway and Virtual Services - but something changed since this config is working in 1.5.

**Comment by jmazzitelli** [comment]
> Agreed. Would you like me to close this issue (again :) )? At this point, it looks like this should be an issue on the istio github if anything.

Yes. I think you will get more exposure from Istio folks if you create an issue over there. Just copy what you said here over there - include your yaml for those resources as well as the weird telemetry.

**Comment by jbilliau5668** [comment]
Sounds good, thanks John.

**Comment by greg-caffeine** [comment]
@jbilliau5668 :  Did you ever get a resolution to this problem?  We are having the same issue and I have not found any other discussion of it.

**Comment by jmazzitelli** [comment]
@greg-caffeine see the associated istio issue - Istio fixed a bug on their end that corrected this - see comment https://github.com/istio/istio/issues/25274#issuecomment-655871226 and following

**Comment by jshaughn** [comment]
Looking at the Istio issue, I think this means the fix would have been in Istio 1.11.  @greg-caffeine are you using Istio version before 1.11?

**Comment by greg-caffeine** [comment]
Yes, we're on 1.9!  Thanks so much for the pointer!
